### ISSUE TITLE
incorrect type for elements in a list

### ISSUE THREAD

**Issue by talkaminker** [PyCQA/pylint#2191]
### Steps to reproduce
pylint the following
```
l = [None]
l[0] = {'abc': 5}
a = l[0]
if 'abc' in a: #this is the line that pylint complains about
    print(1)
```

### Current behavior
E:  4,12: Value 'a' doesn't support membership test (unsupported-membership-test)

### Expected behavior
No error
pylint should undestand that the type has changed to dict (or at least understand that it no longer can guess the type)

### Note:
adding type annotation to the first line like this:
`l: List[Optional[dict]] = [None]`
doesn't fix it

### pylint --version output
No config file found, using default configuration
pylint 1.9.2,
astroid 1.6.5
Python 3.6.5 |Anaconda custom (64-bit)| (default, Mar 29 2018, 13:32:41) [MSC v.1900 64 bit (AMD64)]

**Comment by PCManticore** [comment]
Thanks for the issue! Pylint doesn't yet understand typing annotations, so providing them will not fix this false positive. What I imagine is happening here is that pylint can't see that the value was reassigned later on, since it inference all the potential values that a variable can take during the program execution.

**Comment by nickdrozd** [comment]
The Pylint warning is technically incorrect, but I think its general sentiment that something is wrong here is right. This code happens to work, but it is teetering on the edge of not working. For instance, one line can be added to make it not work:

```python
l = [None]
l[0] = {'abc': 5}
l.insert(0, None)  # <-- whoops!
a = l[0]
if 'abc' in a:
    print(1)
```

I don't think Pylint should endorse code like this. If the code can't be rewritten, `# pylint: disable = unsupported-membership-test` can be added at the offending line.
